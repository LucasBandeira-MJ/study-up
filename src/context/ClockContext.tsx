import { createContext, useEffect, useState } from "react";

let clockTimeout: NodeJS.Timeout

interface ClockContextData {
    time: number
    hours: number
    minutes: number
    seconds: number
    isActive: boolean
    startClock: () => void
    pauseClock: () => void
    stopClock: () => void
}

interface ClockProviderProps {
    children: React.ReactNode
}

export const ClockContext = createContext({} as ClockContextData)

export const ClockProvider = ({children}: ClockProviderProps) => {
    const [time, setTime] = useState<number>(0)
    const [isActive, setIsActive] = useState<boolean>(false)

    useEffect(() => {
        if(isActive) {
            clockTimeout = setTimeout(() => {
                setTime(time => time + 1)
            }, 1000)
        }
    }, [time, isActive])

    const startClock = () => {
        setIsActive(true)
    }

    const pauseClock = () => {
        clearInterval(clockTimeout)
        setIsActive(false)
    }

    const stopClock = () => {
        clearInterval(clockTimeout)
        setIsActive(false)
    }


    const hours = Math.floor(time / 3600)
    const minutes = Math.floor(time / 60 - hours * 60)
    const seconds = Math.floor(time % 60)

    return (
        <ClockContext.Provider value={{
            time,
            hours, 
            minutes, 
            seconds,
            isActive,
            startClock,
            pauseClock,
            stopClock
        }}>
            {children}
        </ClockContext.Provider>
    )
}