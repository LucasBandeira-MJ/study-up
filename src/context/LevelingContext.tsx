import { createContext, ReactNode, useEffect, useState } from "react";

let clockTimeout: NodeJS.Timeout

interface LevelingContextData {
    time: number
    isClockActive: boolean
    showLevelingModal: boolean
    level: number
    currentExperience: number
    experienceToNextLevel: number
    totalExperience: number
    totalLevelUps: number
    handleClock: () => void
    stopClock: () => void
    closeLevelingModal: () => void
}

interface LevelingProviderProps {
    children: ReactNode
}

export const LevelingContext = createContext({} as LevelingContextData)

export const LevelingProvider = ({ children }: LevelingProviderProps) => {
    const [time, setTime] = useState<number>(0)
    const [isClockActive, setIsClockActive] = useState<boolean>(false)
    const [showLevelingModal, setShowLevelingModal] = useState<boolean>(false)
    const [level, setLevel] = useState<number>(1)
    const [currentExperience, setCurrentExperience] = useState<number>(0)
    const [totalExperience, setTotalExperience] = useState<number>(0)
    const [totalLevelUps, setTotalLevelUps] = useState<number>(0)

    const experienceToNextLevel = Math.pow((level + 1) * 2, 2)

    const handleClock = () => {
        if(isClockActive) {
            clearInterval(clockTimeout)
            setIsClockActive(false)
        } else {
            setIsClockActive(true)
        }
    }

    const stopClock = () => {
        clearInterval(clockTimeout)
        setIsClockActive(false)
        setShowLevelingModal(true)
    }

    const closeLevelingModal = () => {
        setShowLevelingModal(false)
        setTime(0)
    }


    const levelUp = () => {
        setLevel(level => level + 1)
        setTotalLevelUps(totalLevelUps => totalLevelUps + 1)
    }


    useEffect(() => {
        if(isClockActive) {
            clockTimeout = setTimeout(() => {
                setTime(time => time + 1)
            }, 1000)
        }
    }, [time, isClockActive])

    useEffect(() => {
        let finalExperience = !!time ?  currentExperience + 1 : currentExperience
        let totalExp = !!time ? totalExperience + 1 : totalExperience

        if(finalExperience >= experienceToNextLevel) {
            finalExperience = finalExperience - experienceToNextLevel
            levelUp()
        }

        setCurrentExperience(finalExperience)
        setTotalExperience(totalExp)
    }, [time])


    return (
        <LevelingContext.Provider value={{
            time,
            isClockActive,
            showLevelingModal,
            level,
            currentExperience,
            experienceToNextLevel,
            totalExperience,
            totalLevelUps,
            handleClock,
            stopClock,
            closeLevelingModal
        }}>
            {children}
        </LevelingContext.Provider>
    )
}

