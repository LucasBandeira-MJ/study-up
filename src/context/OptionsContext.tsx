import { createContext, useState } from "react";
import { OptionsModal } from "../components/OptionsModal";

interface OptionsProviderProps {
    children: React.ReactNode
}

interface OptionsContextData {
    isDarkMode?: boolean
    isOptionsModalOpen: boolean
    toggleTheme: () =>  void
    openOptionsModal: () => void
    closeOptionsModal: () => void
}

export const OptionsContext = createContext({} as OptionsContextData)

export const OptionsProvider = ({children}: OptionsProviderProps) => {
    const [isOptionsModalOpen, setIsOptionsModalOpen] = useState(false)
    const [isDarkMode, setIsDarkMode] = useState(false)

    const toggleTheme = () => {
        setIsDarkMode(theme => !theme)
    }

    const openOptionsModal = () => setIsOptionsModalOpen(true) 
    const closeOptionsModal = () => setIsOptionsModalOpen(false) 

    return (
        <OptionsContext.Provider value={{ 
            isDarkMode, 
            isOptionsModalOpen,
            toggleTheme,
            openOptionsModal,
            closeOptionsModal
        }}>
            {children}
            <OptionsModal />
        </OptionsContext.Provider>   
    )
}