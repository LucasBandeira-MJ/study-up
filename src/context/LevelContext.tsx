import { createContext, useContext, useEffect, useState } from "react";
import { LevelingModal } from "../components/LevelingModal";
import { ClockContext } from "./ClockContext";

interface LevelContextData {
    level: number
    currentExperience: number
    experienceToNextLevel: number
    totalExperience: number
    totalLevelUps: number
}

interface LevelProviderProps {
    children: React.ReactNode
}

export const LevelContext = createContext({} as LevelContextData)

export const LevelProvider = ({children}: LevelProviderProps) => {
    const { time } = useContext(ClockContext)

    const [level, setLevel] = useState<number>(1)
    const [currentExperience, setCurrentExperience] = useState<number>(0)
    const [totalExperience, setTotalExperience] = useState<number>(0)
    const [totalLevelUps, setTotalLevelUps] = useState<number>(0)

    const experienceToNextLevel = Math.pow((level + 1) * 2, 2)

    const levelUp = () => {
        setLevel(level => level + 1)
        setTotalLevelUps(totalLevelUps => totalLevelUps + 1)
    }

    useEffect(() => {
        console.log('uh')
        let finalExperience = !!time ?  currentExperience + 1 : currentExperience
        let totalExp = !! time ? totalExperience + 1 : totalExperience

        if(finalExperience >= experienceToNextLevel) {
            finalExperience = finalExperience - experienceToNextLevel
            levelUp()
        }

        setCurrentExperience(finalExperience)
        setTotalExperience(totalExp)
    }, [time])


    return (
        <LevelContext.Provider value={{
            level,
            currentExperience,
            experienceToNextLevel,
            totalExperience,
            totalLevelUps
        }}>
            {children}
            <LevelingModal />
        </LevelContext.Provider>
    )
}