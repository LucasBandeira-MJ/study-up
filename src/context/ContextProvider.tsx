import { LevelingProvider } from "./LevelingContext"
import { OptionsProvider } from "./OptionsContext"
import { SessionProvider } from 'next-auth/react'

interface ContextProviderProps {
    children: React.ReactNode
}

export const ContextProvider = ({children}: ContextProviderProps) => {
    return (
        <SessionProvider>
            <OptionsProvider>
                <LevelingProvider>
                    {children}
                </LevelingProvider>
            </OptionsProvider>
        </SessionProvider>
    )
}