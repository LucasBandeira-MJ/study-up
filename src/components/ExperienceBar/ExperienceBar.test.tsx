import { render, screen } from "@testing-library/react"
import { ExperienceBar } from "."
import { LevelingContext } from "../../context/LevelingContext"
import renderer from 'react-test-renderer'


describe("ExperienceBar component", () => {
    it("should render correctly", () => {
        render (
            <LevelingContext.Provider value={{ currentExperience: 300, experienceToNextLevel: 1440  } as any }>
                <ExperienceBar />
            </LevelingContext.Provider>
        )

        const initialExperience = screen.queryAllByText(/0 xp/)[0]
        const currentExperience = screen.getByText(/300 xp/)
        const experienceToNextLevel = screen.getByText(/1440 xp/)

        expect(initialExperience).toBeInTheDocument()
        expect(currentExperience).toBeInTheDocument()
        expect(experienceToNextLevel).toBeInTheDocument()
    })

    it('should match the snapshot', () => {
        const snapshot = renderer
            .create(
                <LevelingContext.Provider value={{ currentExperience: 350, experienceToNextLevel: 10440 } as any}>
                    <ExperienceBar />
                </LevelingContext.Provider>
            ).toJSON()

            expect(snapshot).toMatchSnapshot()
      });
})