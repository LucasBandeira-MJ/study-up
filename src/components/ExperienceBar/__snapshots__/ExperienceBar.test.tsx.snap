// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExperienceBar component should match the snapshot 1`] = `
<header
  className="experienceBar"
>
  <span>
    0 xp
  </span>
  <div
    className="experienceContainer"
  >
    <div
      className="experienceFill"
      style={
        Object {
          "width": "3.3524904214559386%",
        }
      }
    />
    <span
      className="currentExperience"
      style={
        Object {
          "left": "3.3524904214559386%",
        }
      }
    >
      350
       xp
    </span>
  </div>
  <span>
    10440
     xp
  </span>
</header>
`;
